pkg/publicshare/manager/memory/memory.go

```diff
@@ -34,50 +37,145 @@ func init() {
 	registry.Register("memory", New)
 }
 
-// New returns a new memory manager
-func New(c map[string]interface{}) (publicshare.Manager, error) {
-	return &manager{}, nil
+// New returns a new memory manager.
+func New() (publicshare.Manager, error) {
+	return &manager{
+		shares: sync.Map{},
+	}, nil
 }
 
-type manager struct{}
+type manager struct {
+	shares sync.Map
+}
 
-// TODO(refs) implement application logic.
-func (m *manager) CreatePublicShare(ctx context.Context, u *authv0alphapb.User, md *storageproviderv0alphapb.ResourceInfo, g *publicshareproviderv0alphapb.Grant) (*publicshareproviderv0alphapb.PublicShare, error) {
-	return &publicshareproviderv0alphapb.PublicShare{}, nil
+// CreatePublicShare safely adds a new entry to manager.shares
+func (m *manager) CreatePublicShare(ctx context.Context, u *authv0alphapb.User, rInfo *storageproviderv0alphapb.ResourceInfo, g *publicshareproviderv0alphapb.Grant) (*publicshareproviderv0alphapb.PublicShare, error) {
+	// what to set id?
+	//  TODO does OpaqueId expecting a specific format?
+	// where could this initialization go wrong and early return?
+	id := &publicshareproviderv0alphapb.PublicShareId{
+		OpaqueId: randString(12),
+	}
+	tkn := randString(12)
+	now := uint64(time.Now().Unix())
+
+	newShare := publicshareproviderv0alphapb.PublicShare{
+		Id:          id,
+		Owner:       u.Id,
+		Creator:     u.Id,
```

Comment:
Owner vs Creator: https://cs3org.github.io/cs3apis/#cs3.publicshareproviderv0alpha.PublicShare
Owner = Uniquely identifies the owner of the share (the resource owner at the time of creating the share).
Creator = Uniquely identifies a principal who initiates the share creation.

Reply:
perfect!